File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_updates/lib/src/state/updates_state.dart
import '../models/updates_models.dart';

enum UpdatesStatus { idle, loading, ready, error }

class UpdatesState {
  const UpdatesState({
    this.status = UpdatesStatus.idle,
    this.lastResult,
    this.error,
  });

  final UpdatesStatus status;
  final UpdatesPageResult<Object?>? lastResult;
  final String? error;

  UpdatesState loading() => const UpdatesState(status: UpdatesStatus.loading);

  UpdatesState loaded(UpdatesPageResult<Object?> result) =>
      UpdatesState(status: UpdatesStatus.ready, lastResult: result);

  UpdatesState failed(String message) =>
      UpdatesState(status: UpdatesStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-updates/src/state/updatesState.ts
import type { UpdatesPageResult } from "../types/updatesModels.js";

export interface UpdatesState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: UpdatesPageResult<unknown>;
  readonly error?: string;
}

export function initialUpdatesState(): UpdatesState {
  return { status: "idle" };
}

export function reduceUpdatesState(
  state: UpdatesState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: UpdatesPageResult<unknown> }
    | { type: "failed"; error: string },
): UpdatesState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
